lib/methods/ifwapichecktask.cpp

```diff
@@ -0,0 +1,467 @@
+		);
+		return;
+	}
+
+	if (!sslConn.IsVerifyOK()) {
+		auto cert (sslConn.GetPeerCertificate());
+		Value cn;
+
+		try {
+			cn = GetCertificateCN(cert);
+		} catch (const std::exception&) {
+		}
+
+		ReportIfwCheckResult(
+			yc, checkable, cmdLine, cr,
+			"Certificate validation failed for IfW API on host '" + psHost + "' (SNI: '" + san + "'; CN: "
+				+ (cn.IsString() ? "'" + cn + "'" : "N/A") + ") port '" + psPort + "': " + sslConn.GetVerifyError(),
+			start
+		);
+		return;
+	}
+
+	try {
+		async_write(conn, req, yc);
+		conn.async_flush(yc);
+	} catch (const std::exception& ex) {
+		ReportIfwCheckResult(
+			yc, checkable, cmdLine, cr,
+			"Can't send HTTP request to IfW API on host '" + psHost + "' port '" + psPort + "': " + ex.what(), start
+		);
+		return;
+	}
+
+	try {
+		async_read(conn, buf, resp, yc);
+	} catch (const std::exception& ex) {
+		ReportIfwCheckResult(
+			yc, checkable, cmdLine, cr,
+			"Can't read HTTP response from IfW API on host '" + psHost + "' port '" + psPort + "': " + ex.what(), start
+		);
+		return;
+	}
+
+	double end = Utility::GetTime();
+
+	{
+		boost::system::error_code ec;
+		sslConn.async_shutdown(yc[ec]);
+	}
+
+	CpuBoundWork cbw (yc);
+	Value jsonRoot;
+
+	try {
+		jsonRoot = JsonDecode(resp.body());
+	} catch (const std::exception& ex) {
+		ReportIfwCheckResult(
+			checkable, cmdLine, cr,
+			"Got bad JSON from IfW API on host '" + psHost + "' port '" + psPort + "': " + ex.what(), start, end
+		);
+		return;
+	}
+
+	if (!jsonRoot.IsObjectType<Dictionary>()) {
+		ReportIfwCheckResult(
+			checkable, cmdLine, cr,
+			"Got JSON, but not an object, from IfW API on host '" + psHost + "' port '" + psPort + "'", start, end
+		);
+		return;
+	}
+
+	Value jsonBranch;
+
+	if (!Dictionary::Ptr(jsonRoot)->Get(psCommand, &jsonBranch)) {
+		ReportIfwCheckResult(
+			checkable, cmdLine, cr,
+			"Missing ." + psCommand + " in JSON object from IfW API on host '" + psHost + "' port '" + psPort + "'", start, end
+		);
+		return;
+	}
+
+	if (!jsonBranch.IsObjectType<Dictionary>()) {
+		ReportIfwCheckResult(
+			checkable, cmdLine, cr,
+			"." + psCommand + " is not an object in JSON from IfW API on host '" + psHost + "' port '" + psPort + "'", start, end
+		);
+		return;
+	}
+
+	Dictionary::Ptr result = jsonBranch;
+	double exitcode;
+
+	try {
+		exitcode = result->Get("exitcode");
+	} catch (const std::exception& ex) {
+		ReportIfwCheckResult(
+			checkable, cmdLine, cr,
+			"Got bad exitcode from IfW API on host '" + psHost + "' port '" + psPort + "': " + ex.what(), start, end
+		);
+		return;
+	}
+
+	Array::Ptr perfdata;
+
+	try {
+		perfdata = result->Get("perfdata");
+	} catch (const std::exception& ex) {
+		ReportIfwCheckResult(
+			checkable, cmdLine, cr,
+			"Got bad perfdata from IfW API on host '" + psHost + "' port '" + psPort + "': " + ex.what(), start, end
+		);
+		return;
+	}
+
+	ReportIfwCheckResult(checkable, cmdLine, cr, result->Get("checkresult"), start, end, exitcode, perfdata);
```

Comment:
Insufficient input validation:
- `exitcode` can be any `double` value. If the response JSON contains `"exitcode": 42.23`, this makes it to the `/v1/objects` API response as `"state": 42` and is also written to Redis by the Icinga DB feature this way, breaking Icinga DB Web ("Invalid service state 42" and a stack trace on all pages showing the service (including the service list of the host for example)).
- `perfdata` can be pretty much any array. If the response JSON contains `"perfdata": [ 42, {"will it break?": "let's see!"} ]`, this makes it to the `/v1/objects` API like this:
    ```json
          "performance_data": [
            42,
            {
              "counter": false,
              "crit": null,
              "label": "",
              "max": null,
              "min": null,
              "type": "PerfdataValue",
              "unit": "",
              "value": 0,
              "warn": null
            }
          ],
    ```
  I think this may also throw a surprise exception in `perfdata->Join(" ")`, but I haven't explicitly tested that branch so far.
- `result->Get("checkresult")` can also be pretty much anything. As it's passed as a `const String&` argument, it's implicitly converted by `Value::operator String()` which happens to be able to convert anything, but the code here doesn't really look robust.

In general, the code should perform more strict validation on the input, in particular, the exit code and perfdata should be validated in a similar way that is done for `exec()` checks. Also it wouldn't hurt to just enforce that things are represented as numbers/strings in the JSON, not just something that the `Value` class happens to be able to convert.

I used the following for testing (needs to be started manually just before each check is started):

```sh
(sleep 10; cat response.txt) | openssl s_server -cert master-1.crt -key master-1.key
```

<details>
<summary>response.txt (save with DOS line endings!)</summary>

```
HTTP/1.1 200 OK
Server: localhost
Content-Type: application/json
Content-Length: 148

{ 
  "dummy": {
    "exitcode": 42.23,
    "checkresult": "fun with perfdata",
    "perfdata": [ 42, {"will it break?": "let's see!"} ]
  }
}
```
</details>

Reply:
Sure. But we also don’t check [0, 2^64) for every uint64 on Icinga DB side, do we? No, **the input** has not to be garbage in the first place.

Reply:
Do we really have to discuss this? The code is passing on garbage and **things do break because of this**:
- Icinga DB Web breaks down because of this invalid value.
- There is code that assumes the state to be in some range: https://github.com/Icinga/icinga2/blob/eddd4c7bf7554a3700abe880781a7646703b7e5f/lib/icinga/checkable-check.cpp#L285-L295
- There even is code that uses `VERIFY(...)` (i.e. assert also present in release builds) to ensure the state is in the valid range: https://github.com/Icinga/icinga2/blob/eddd4c7bf7554a3700abe880781a7646703b7e5f/lib/icinga/notification.cpp#L668-L694
 **This is bad.**

Reply:
> * `result->Get("checkresult")` can also be pretty much anything. As it's passed as a `const String&` argument, it's implicitly converted by `Value::operator String()` which happens to be able to convert anything, but the code here doesn't really look robust.
> 
> In general, the code should perform more strict validation on the input, in particular, the exit code and perfdata should be validated in a similar way that is done for `exec()` checks. Also it wouldn't hurt to just enforce that things are represented as numbers/strings in the JSON, not just something that the `Value` class happens to be able to convert.

The just pushed ones allow no/null perfdata and still any type of checkresult. Strictly speaking it's indeed not _really robust_ – it's _einfach geil™️_ 😎

## TODO

* [ ] opinion @julianbrost
* [ ] (possibly even unit)tests @Al2Klimov

Reply:
> The just pushed ones

You haven't pushed anything?

> Strictly speaking it's indeed not _really robust_ – it's _einfach geiltm_ sunglasses

But now I'm scared.